addons/mail/controllers/discuss_controller.py

```diff
@@ -0,0 +1,307 @@
+        # If it's an internal user, there is no further restrictions.
+        if not request.env.user.share:
+            return attachment.sudo(False).unlink()
+
+        message = attachment.env['mail.message'].search([('attachment_ids', 'in', attachment.ids)], limit=1)
+        if not message: # the attachment has not been sent yet
+            # An access token is provided to guests so they can delete their
+            # own pending attachments. Ensure this token is valid:
+            if not consteq(access_token, attachment.access_token):
+                raise UserError(_("The attachment does not exist or you do not have the rights to access it."))
+
+            # Ensure the attachment is in a pending state.
+            if attachment.res_model != 'mail.compose.message' or attachment.res_id != 0:
+                raise UserError(_("The attachment cannot be removed because it is not in a pending state."))
+        else:
+            # If it's a portal user or a guest, ensure they are the uploader
+            # of the attachment. The uploader of an attachment is determined
+            # from the author of the message to which the attachment is linked.
+            error_message = _("You cannot delete an attachment that has not been uploaded by you.")
+            # portal user
+            if request.session.uid and (not message.author_id or message.author_id != request.env.user.partner_id):
+                raise UserError(error_message)
+            # guest
+            if not request.session.uid and (not message.guest_id or message.guest_id != request.env['mail.guest']._get_guest_from_request(request)):
+                raise UserError(error_message)
+
+        return attachment.unlink()
+
+    @http.route('/mail/attachment/upload', type='http', auth='public', methods=['POST'], website=True)
+    def upload_attachment(self, file_to_upload, thread_id, thread_model, **kwargs):
+        channel_partner = request.env['mail.channel.partner']
+        # Files can be uploaded by users. For guests, they can only upload to 
+        # a channel, and only if they are members of it.
+        if thread_model == 'mail.channel':
+            channel_partner = self._get_channel_partner(int(thread_id))
+        if not channel_partner and request.env.user.share:
+            raise NotFound()
+
+        # Only generate an access token if the current user is a guest. Also,
+        # use the env of channel_partner so that the guest is defined in the 
+        # context and CRUD are performed in sudo mode.
+        IrAttachment = request.env['ir.attachment']
+        access_token = False
+        if request.env.user.share:
+            IrAttachment = channel_partner.env['ir.attachment']
+            access_token = IrAttachment._generate_access_token()
+        
+        attachment = IrAttachment.create({
+            'access_token': access_token,
+            'datas': base64.b64encode(file_to_upload.read()),
```

Comment:
@seb-odoo a while ago irattachment grew a `raw` member which gives access to the raw payload of ir.attachment, rather than the base64-encoded `datas`.

So if you need or have base64 data it's fine to use `datas`, but if you have binary data (as here) all it does is add unnecessary overhead as the payload gets encoded, then immediately decoded, before being stored in the attachment store.